software/tests/experimental/test_knobs.py

```diff
@@ -0,0 +1,263 @@
+    # Unlocked
+    assert locked_knob.state == LockableKnob.STATE_UNLOCKED
+
+    locked_knob.request_unlock()
+    assert locked_knob.state == LockableKnob.STATE_UNLOCKED
+
+    locked_knob.lock()
+    assert locked_knob.state == LockableKnob.STATE_LOCKED
+
+    # locked
+    locked_knob.lock()
+    assert locked_knob.state == LockableKnob.STATE_LOCKED
+
+    locked_knob.request_unlock()
+    assert locked_knob.state == LockableKnob.STATE_UNLOCK_REQUESTED
+
+    # Unlock requested
+
+    locked_knob.request_unlock()
+    assert locked_knob.state == LockableKnob.STATE_UNLOCK_REQUESTED
+
+    locked_knob.lock()
+    assert locked_knob.state == LockableKnob.STATE_LOCKED
+
+
+def test_access_percent(mockHardware: MockHardware, locked_knob: LockableKnob):
+    assert round(locked_knob.percent(), 2) == 0.50
+
+    mockHardware.set_ADC_u16_value(locked_knob, MAX_UINT16)
+    assert round(locked_knob.percent(), 2) == 0
+
+    mockHardware.set_ADC_u16_value(locked_knob, 0)
+    assert round(locked_knob.percent(), 2) == 1
+
+
+def test_access_choice(mockHardware: MockHardware, locked_knob: LockableKnob):
+    assert locked_knob.choice([1, 2, 3]) == 2
+
+    mockHardware.set_ADC_u16_value(locked_knob, MAX_UINT16)
+    assert locked_knob.choice([1, 2, 3]) == 1
+
+    mockHardware.set_ADC_u16_value(locked_knob, 0)
+    assert locked_knob.choice([1, 2, 3]) == 3
+
+
+def test_access_range(mockHardware: MockHardware, locked_knob: LockableKnob):
+    assert locked_knob.range() == 49
+
+    mockHardware.set_ADC_u16_value(locked_knob, MAX_UINT16)
+    assert locked_knob.range() == 0
+
+    mockHardware.set_ADC_u16_value(locked_knob, 0)
+    assert locked_knob.range() == 99
+
+
+def test_access_read_position(mockHardware: MockHardware, locked_knob: LockableKnob):
+    assert locked_knob.read_position() == 49
+
+    mockHardware.set_ADC_u16_value(locked_knob, MAX_UINT16)
+    assert locked_knob.read_position() == 0
+
+    mockHardware.set_ADC_u16_value(locked_knob, 0)
+    assert locked_knob.read_position() == 99
+
+
+# Disabled Knob Tests
+
+
+# KnobBank tests
+
+
+@pytest.fixture
+def knob_bank(mockHardware: MockHardware):
+    kb = (
+        KnobBank.builder(k1)
+        .with_disabled_knob()
+        .with_unlocked_knob("param1")
+        .with_locked_knob("param2", MAX_UINT16)
+        .build()
+    )
+
+    # start the knob in the middle
+    mockHardware.set_ADC_u16_value(k1, MAX_UINT16 / 2)
+
+    return kb
+
+
+def test_next_mode(knob_bank: KnobBank):
+    assert knob_bank.index == 1
+    knob_bank.next()
+    assert knob_bank.index == 2
+    knob_bank.next()
+    assert knob_bank.index == 0
+    knob_bank.next()
+    assert knob_bank.index == 1
+
+
+def test_access_by_name(mockHardware: MockHardware, knob_bank: KnobBank):
+    # knob starts in the middle, knob 1
+    assert round(knob_bank.param1.percent(), 2) == 0.50
+    assert round(knob_bank.param2.percent(), 2) == 0
+    assert round(knob_bank.current.percent(), 2) == 0.50
```

Comment:
This is a really nice interface for dealing with several parameters controlled by a single knob. Well done!

Reply:
Thank you!